### PR TITLE
WIP: fix - workspaces and appbar respect titlebar size and position

### PR DESCRIPTION
**Pull request by ramirezmike**

### CODE REVIEW

File: src/display.rs [hunk 1]
```diff
@@ -1,16 +1,53 @@
...
+#[derive(Default, Debug, Clone, Copy)]
+pub struct TaskBar {
+    pub hwnd: i32,
+    pub height: i32,
+    pub width: i32,
+    pub position: TaskBarPosition
+}
+
+lazy_static! {
+    pub static ref CURRENT_INITIALIZING_DISPLAY: Mutex<Option<Display>> = Mutex::new(None);
```
Comment (TimUntersberger) [inline src/display.rs:49]: If possible I'd like to avoid static variables.

File: src/display.rs [hunk 2]
```diff
@@ -77,19 +159,121 @@ pub fn init() {
...
+
+    sync_task_bars();
+}
+
+pub fn sync_task_bars() {
     let mut displays = DISPLAYS.lock().unwrap();
+    for display in displays.iter_mut() {
+        *display = sync_task_bar(*display);
+    }
+}
```
Comment (TimUntersberger) [inline src/display.rs:184]: I don't think that the `Display` should care about synchronization.

**Comment by TimUntersberger** [comment]
> 
> 
> Perhaps the issue is the show/hide should only be called on the primary taskbar? Just a note for myself to check when I get back to this..

> I fixed this before. If you merge the latest changes from development you should have the fix.
>
> TLDR: You just have to hide the primary monitor taskbar first

File: src/display.rs (hunk 2 above)
Comment (TimUntersberger) [reply src/display.rs:184]: Maybe there is an event that you can use in the `win_event_handler`?

File: src/display.rs [hunk 3]
```diff
@@ -21,14 +58,59 @@ pub struct Display {
     pub right: i32,
     pub top: i32,
     pub bottom: i32,
+    pub task_bar: Option<TaskBar>
```
[the remark this answers is not in the document]
Comment (ramirezmike) [reply src/display.rs:61]: None when the taskbar doesn't exist on the display. When it's Some it exists on the display at a specific position but could be hidden.

[the remark this answers is not in the document]
Comment (ramirezmike) [reply src/display.rs:61]: Actually, for this I'll add a hidden position and use that to eliminate the calls to IsWindowVisible to only when updating the taskbar

File: src/display.rs (hunk 2 above)
Comment (ramirezmike) [reply src/display.rs:184]: "sync" may have been a poor choice. This is to update the state of the taskbars on each display after the user does something like update the config or toggle the taskbar. I can rename this.

As far as win_event_handler, that will likely be how to handle if the user moves the taskbar or goes into the taskbar settings and alters things and then it'll call this function to update the taskbar state. Currently though this function is used to update state based on wwm-triggered taskbar changes.

Comment (TimUntersberger) [reply src/display.rs:184]: > "sync" may have been a poor choice. This is to update the state of the taskbars on each display after the user does something like update the config or toggle the taskbar. I can rename this.

If possible please rename these, thank you 👍 .

File: src/display.rs [hunk 4]
```diff
@@ -21,14 +58,59 @@ pub struct Display {
...
+        let task_bar_height = 
+            if self.task_bar.is_some() && unsafe { IsWindowVisible(self.task_bar.unwrap().hwnd as HWND) == 1 } {
+                let task_bar = self.task_bar.unwrap();
+                match task_bar.position {
+                    TaskBarPosition::Top | TaskBarPosition::Bottom => task_bar.height,
+                    _ => 0
+                }
+            } else { 
+                0 
+            };
```
[the remark this answers is not in the document]
Comment (ramirezmike) [reply src/display.rs:75]: I agree with you in principle, the only argument I have for using this implementation is if you view the Display struct as a "working area" of a monitor instead of as a 1-to-1 representation to the monitor.

Treating it as a monitor representation would mean when retrieving the width the user would also have to handle subtracting the taskbar size if the taskbar is on the left/right and the same for height and top/bottom. Without this, you'd get overlap of the tilegrids, the appbar and the positions of items in the appbar would also be offset.

That said, I'm not against changing it. Would one of the following work?

A) have width/height take a parameter to factor in taskbar
B) have a separate set of functions to get "workspace" width/height
C) have separate functions to get taskbar sizes and update references to display.width to subtract the taskbar's width manually as needed? That is something like
    display.width() - display.task_bar_width()

File: src/display.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimUntersberger) [reply src/display.rs:61]: > None when the taskbar doesn't exist on the display. When it's Some it exists on the display at a specific position but could be hidden.

Could you please tell me an example where the taskbar doesn't exist on a display?

[the remark this answers is not in the document]
Comment (ramirezmike) [reply src/display.rs:61]: when you have multiple monitors and you uncheck the taskbar setting "Show taskbar on all displays". If you do that, then the primary display has a taskbar but every other display doesn't. That's what None is used for, otherwise I store a reference to the taskbar for showing/hiding.

File: src/display.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (TimUntersberger) [reply src/display.rs:75]: > have a separate set of functions to get "workspace" width/height

👍. This would even align with the winapi, maybe we can even use terminology from there. I think they used `working area`.

**Comment by ramirezmike** [comment]
I've made the changes we discussed above and everything worked well until I tried having a taskbar on multiple screens but with different positions: top on primary and right side on secondary screen. It turns out that ABM_GETTASKBARPOS only works for the taskbar on the primary screen. Which, everything works when the taskbars are in the same positions on all screens but obviously doesn't work otherwise.

So, I looked to see what else I could do and saw (this post)[https://devblogs.microsoft.com/oldnewthing/20141218-00/?p=43353] which suggests just getting the work area. So I made some changes to do things that way instead. Having a busy week but should have more time in a few days.

**Comment by TimUntersberger** [comment]
Can't you just calculate the position yourself? Just get the `RECT` of the taskbar window and based on left, right, top and bottom you know where it is. The easiest way would be to just save a `Window` struct in the `Taskbar` struct, it already has a method for this.

File: src/bar.rs [hunk 5]
```diff
@@ -239,14 +239,14 @@ unsafe extern "system" fn window_cb(
             .unwrap();
 
         if CLEAR.load(Ordering::SeqCst) {
-            clear_section(hdc, height, 0, display.width());
+            clear_section(hdc, height, 0, display.working_area_width());
```
Comment (ramirezmike) [inline src/bar.rs:242]: there were a couple of these I had to update to use the working_area_width because it was causing overlaps between the bar and the taskbar if you have it on the left or right side

File: src/display.rs [hunk 6]
```diff
@@ -30,6 +27,56 @@ impl Display {
     pub fn width(&self) -> i32 {
         self.right - self.left
     }
+    pub fn working_area_height(&self) -> i32 {
```
Comment (ramirezmike) [inline src/display.rs:30]: this group of functions haven't changed much from the previous version, just references to task_bar::

File: src/hot_reload.rs [hunk 7]
```diff
@@ -28,25 +41,22 @@ pub fn update_config(new_config: Config) -> Result<(), Box<dyn std::error::Error
...
             draw_app_bar = true;
 
             for d in DISPLAYS.lock().unwrap().iter_mut() {
                 d.bottom -= config.app_bar_height;
             }
 
+            update_grid_displays = true;
+        }
+
+        if update_grid_displays {
```
Comment (ramirezmike) [inline src/hot_reload.rs:55]: I moved this block of code here and used `update_grid_displays` to determine if it needs to get run since I ended up with this block in four places.

File: src/task_bar.rs [hunk 8]
```diff
@@ -1,93 +1,134 @@
...
-            return 0;
-        }
-    }
+#[derive(Debug, Clone, Copy)]
+pub enum TaskBarPosition {
+    Top,
+    Bottom,
+    Left,
+    Right,
+    Hidden
```
Comment (ramirezmike) [inline src/task_bar.rs:29]: Again, `Hidden` is for when the task_bar exists but has been hidden by the user as opposed to it not existing on the display which is when the display's reference to `Option<TaskBar>` is `None`

File: src/task_bar.rs [hunk 9]
```diff
@@ -1,93 +1,134 @@
...
+    foreach_taskbar(|hwnd| {
+        info!("Hiding taskbar {}", hwnd);
+        unsafe {
+            ShowWindow(hwnd as HWND, SW_HIDE);
+        }
+    });
 
+    update_task_bars();
+}
 fn foreach_taskbar(cb: fn(i32) -> ()) {
```
Comment (ramirezmike) [inline src/task_bar.rs:64]: this section is modified slightly but in effect does the same thing except uses `Display` references, which have a reference to their respective `task_bar`. It's the same sort to make sure the `task_bars` get hidden/shown in the right order to make sure they're all hidden or shown.

One thing to note: if the user has their config `multi_monitor` set to false _BUT_ has a windows task_bar on each screen, this will only affect the task_bar on the Primary display. That is, if the user has the task_bar on every display but is running Nog on only one display, the task_bar will not be affected on the other displays. That seemed like the sensible way to handle this.

File: src/task_bar.rs [hunk 10]
```diff
@@ -1,93 +1,134 @@
...
+        EnumWindows(Some(enum_windows_cb), 0);
     }
 }
 
-pub fn hide() {
-    foreach_taskbar(|hwnd| {
-        info!("Hiding taskbar {}", hwnd);
-        unsafe {
-            ShowWindow(hwnd as HWND, SW_HIDE);
+unsafe extern "system" fn enum_windows_cb(hwnd: HWND, _: LPARAM) -> BOOL {
```
Comment (ramirezmike) [inline src/task_bar.rs:83]: this function iterates over the windows, finds the taskbars and adds them to their respective displays if they're "managed" by nog and returns early if it hits a primary display and multi_monitor is false.

File: src/task_bar.rs [hunk 11]
```diff
@@ -1,93 +1,134 @@
...
+        if !CONFIG.lock().unwrap().multi_monitor && is_display_primary {
+            return 0;
         }
-    });
+    }
+
+    1
+}
+
+fn get_taskbar_position(rect: RECT, hwnd: HWND, hmonitor: i32) -> TaskBarPosition {
```
Comment (ramirezmike) [inline src/task_bar.rs:121]: this should be pretty self explanatory but all the task_bar position detection is handled in this function. It gets the rect of the display rather than using Display because I found that the Display's bottom gets modified if the app_bar is visible. 

I didn't want to mess around too much with that aspect of the app_bar/display relationship in case you want to have something in the future that allows the user to put the app_bar in another location.

**Comment by ramirezmike** [comment]
I ended up just using the taskbar rect to calculate. This works well for me testing different configurations on my two monitors but I don't have a third one to test out on.

But this resolved the issue for me, showing/hiding taskbars and appbars in different positions as well as taskbars in different positions on different displays. Let me know what you think.

Also, unrelated, I'm really digging the config changes. Good stuff

**Comment by ramirezmike** [comment]
I thought about this a bit more and think it can be optimized a bit if I update the taskbars by iterating over the displays instead of doing the enumwindows when the user toggles visibility of the taskbar. I haven't tried it yet, just realized that might be a better approach.

File: src/display.rs [hunk 12]
```diff
@@ -30,6 +27,56 @@ impl Display {
     pub fn width(&self) -> i32 {
         self.right - self.left
     }
+    pub fn working_area_height(&self) -> i32 {
+        let task_bar_height = 
+            if self.task_bar.is_some() {
```
Comment (TimUntersberger) [inline src/display.rs:32]: Is there a reason you are not using this?
```rust
if let Some(task_bar) = self.task_bar { ... }
```

File: src/display.rs [hunk 13]
```diff
@@ -30,6 +27,56 @@ impl Display {
...
+                }
+            } else { 
+                0 
+            };
+
+        self.height() - task_bar_height
+    }
+    pub fn working_area_width(&self) -> i32 {
+        let task_bar_width =
+            if self.task_bar.is_some() {
```
Comment (TimUntersberger) [inline src/display.rs:46]: Same as above

File: src/display.rs [hunk 14]
```diff
@@ -30,6 +27,56 @@ impl Display {
...
+                    _ => 0
+                }
+            } else { 
+                0 
+            };
+
+        self.width() - task_bar_width
+    }
+    pub fn vertical_offset(&self) -> i32 {
+        if self.task_bar.is_some() {
```
Comment (TimUntersberger) [inline src/display.rs:59]: Same as above

File: src/display.rs [hunk 15]
```diff
@@ -30,6 +27,56 @@ impl Display {
...
+            match task_bar.position {
+                task_bar::TaskBarPosition::Top => task_bar.height,
+                _ => 0
+            }
+        } else { 
+            0 
+        }
+    }
+    pub fn horizontal_offset(&self) -> i32 {
+        if self.task_bar.is_some() {
```
Comment (TimUntersberger) [inline src/display.rs:70]: Same as above

File: src/display.rs [hunk 16]
```diff
@@ -30,6 +27,56 @@ impl Display {
...
+            let task_bar = self.task_bar.unwrap();
+            match task_bar.position {
+                task_bar::TaskBarPosition::Top => task_bar.height,
+                _ => 0
+            }
+        } else { 
+            0 
+        }
+    }
+    pub fn horizontal_offset(&self) -> i32 {
```
Comment (TimUntersberger) [inline src/display.rs:69]: I think this name is a bit confusing. If I read `horizontal_offset` I don't really know whether the offset is on the left or right side.

File: src/display.rs [hunk 17]
```diff
@@ -30,6 +27,56 @@ impl Display {
...
+                    task_bar::TaskBarPosition::Left | task_bar::TaskBarPosition::Right => task_bar.width,
+                    _ => 0
+                }
+            } else { 
+                0 
+            };
+
+        self.width() - task_bar_width
+    }
+    pub fn vertical_offset(&self) -> i32 {
```
Comment (TimUntersberger) [inline src/display.rs:58]: I think this name is a bit confusing. If I read `vertical_offset` I don't really know whether the offset is on the top or bottom side.

File: src/task_bar.rs [hunk 18]
```diff
@@ -1,93 +1,134 @@
...
+
+        let hmonitor = MonitorFromWindow(hwnd, MONITOR_DEFAULTTONULL) as i32;
+        let task_bar = TaskBar {
+            hwnd: hwnd as i32,
+            height: rect.bottom - rect.top,
+            width: rect.right - rect.left,
+            position: get_taskbar_position(rect, hwnd, hmonitor)
+        };
+
+        debug!("Initialized Taskbar(hwnd: {}, height: {}, width: {}, position: {:?})",
```
Comment (TimUntersberger) [inline src/task_bar.rs:102]: Can you extract `Taskbar(hwnd: {}, height: {}, width: {}, position: {:?})` into a `Debug` implementation for the `TaskBar` struct?

File: src/task_bar.rs [hunk 19]
```diff
@@ -1,93 +1,134 @@
...
+    1
+}
+
+fn get_taskbar_position(rect: RECT, hwnd: HWND, hmonitor: i32) -> TaskBarPosition {
+    let mut monitor_info = MONITORINFO { cbSize: size_of::<MONITORINFO>() as u32, ..MONITORINFO::default() };
+    unsafe {
+        GetMonitorInfoA(hmonitor as HMONITOR, &mut monitor_info);
+    }
+    let RECT { left, right, top, bottom } = monitor_info.rcMonitor;
+    let is_window_visible = unsafe { IsWindowVisible(hwnd) == 1 };
```
Comment (TimUntersberger) [inline src/task_bar.rs:127]: If I remember correctly this should be true 100% of the time. If you have the taskbar set to auto hide in Windows it is still visible, but only 2px big.

File: src/task_bar.rs (hunk 11 above)
Comment (TimUntersberger) [reply src/task_bar.rs:121]: > It gets the rect of the display rather than using Display because I found that the Display's bottom gets modified if the app_bar is visible.

This is actually the same problem as with the taskbar. I don't think that the bar should influence the display size, but this should probably be in a seperate PR as this influences other thing too.

File: src/task_bar.rs (hunk 19 above)
Comment (ramirezmike) [reply src/task_bar.rs:127]: The case I found when it wasn't visible is when you have multiple monitors and uncheck the setting to display a taskbar on all monitors. I don't know how often users would change these settings while running Nog, but one scenario that this might come up is when the user plugs/unplugs or docks their computer and it changes all the displays. It's possible that would introduce a display with an "invisible" taskbar. 

While this code isn't handling the display change events yet, I think this function would be what you'd want to call to handle taskbar changes when those events occur.

File: src/display.rs (hunk 12 above)
Comment (ramirezmike) [reply src/display.rs:32]: I'm not sure, I did have that at some point but must have changed it at some point. It makes sense, I've updated these.

File: src/display.rs (hunk 16 above)
Comment (ramirezmike) [reply src/display.rs:69]: I might change these to instead be working_area_left() and working_area_top()

File: src/bar/create.rs [hunk 20]
```diff
@@ -13,13 +13,8 @@ pub fn create() -> Result<(), util::WinApiResultError> {
 
     let name = "nog_bar";
 
-    let mut height_guard = HEIGHT.lock().unwrap();
```
Comment (ramirezmike) [inline src/bar/create.rs:16]: I was resolving merge conflicts and noticed this again and I believe it's something I'm unfamiliar with

```
    let mut height_guard = HEIGHT.lock().unwrap();

    let app_bar_bg = CONFIG.lock().unwrap().app_bar_bg;

    *height_guard = CONFIG.lock().unwrap().app_bar_height;

    let height = *height_guard;
```

A) HEIGHT comes from the task_bar which before my changes would just be the height of the taskbar from the primary display. However, I guess it's technically possible that each display has a different height (why would someone do this??) in which case the appbar for a given display should reference that display's taskbar to get the height. This maybe should be a separate issue but...

B) heightguard ends up getting overwritten to be app_bar_height and then its value gets set to height, right? 

This looks like a pattern I'm not familiar with, what exactly is this "guard" doing? What is the scenario where height doesn't end up as app_bar_height?

Comment (TimUntersberger) [reply src/bar/create.rs:16]: > This looks like a pattern I'm not familiar with, what exactly is this "guard" doing? What is the scenario where height doesn't end up as app_bar_height?

This is probably still left over from a refactor. `guard` is talking about a [MutexGuard](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html) ([Mutex](https://doc.rust-lang.org/std/sync/struct.Mutex.html) is also relevant). If the `height_guard` doesn't get used later on, then you can clean this up if you want to.

> However, I guess it's technically possible that each display has a different height

Are you sure that you can configure this in the windows settings? If there is no official way, then I don't really care whether we support this.

Comment (ramirezmike) [reply src/bar/create.rs:16]: >Are you sure that you can configure this in the windows settings? If there is no official way, then I don't really care whether we support this.

It's technically possible... A user can have multiple monitors with a taskbar on each and can just manually resize one (if they don't have taskbars locked) and end up with different sized taskbar per display but.. again I don't know why someone would do that. It may be overkill to handle.

Comment (TimUntersberger) [reply src/bar/create.rs:16]: Lets ignore this. If someone really needs it, then he will open an Issue anyway.

**Comment by TimUntersberger** [comment]
LGTM. I'll merge this, if you are happy with it.

**Comment by ramirezmike** [comment]
Sounds good to me

**Comment by TimUntersberger** [comment]
Wait why aren't you listed as author? Did I miss something?

**Comment by ramirezmike** [comment]
The tree looks right to me, it shows two commits by me and two merges of those commits by you, it's just the commit message on the merge is different. I don't know why that is but the code content looks correct.
